### PR TITLE
Add experimental dynamic in-context learning variant type to gateway

### PR DESCRIPTION
**Pull request by virajmehta**

### CODE REVIEW

File: gateway/src/variant/dicl.rs [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+        clients: &'request InferenceClients<'request>,
+        inference_params: InferenceParams,
+    ) -> Result<InferenceResult, Error> {
+        let mut inference_params = inference_params;
+        let serialized_input = serde_json::to_string(&input).map_err(|e| Error::Serialization {
+            message: format!(
+                "Error in serializing Input in dynamic in-context learning variant: {}",
+                e
+            ),
+        })?;
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:72]: Only used in function below so do this in function below.

File: gateway/src/variant/dicl.rs [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+    fn get_all_template_paths(&self) -> Vec<&PathBuf> {
+        vec![]
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct ChatExample {
+    input: Input,
+    output: Vec<ContentBlockOutput>,
+    // distance: Option<f32>, // Include if needed
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:227]: ?

File: gateway/src/variant/dicl.rs [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+}
+
+// Start of Selection
+#[derive(Debug, Deserialize, PartialEq)]
+struct JsonExample {
+    input: Input,
+    output: JsonInferenceOutput,
+    // distance: Option<f32>, // Include if needed
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:235]: ?

File: gateway/src/variant/dicl.rs [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct ChatExample {
+    input: Input,
+    output: Vec<ContentBlockOutput>,
+    // distance: Option<f32>, // Include if needed
+}
+
+// Start of Selection
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:230]: ?

File: gateway/src/variant/dicl.rs [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+    Chat(ChatExample),
+    Json(JsonExample),
+}
+
+#[derive(Clone, Debug, Deserialize)]
+struct RawExample {
+    input: String,
+    output: String,
+    // Include `distance` if you need it later
+    // distance: f32,
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:250]: ?

File: gateway/src/variant/dicl.rs [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+        if let Ok(output) = serde_json::from_str::<Vec<ContentBlockOutput>>(&raw_example.output) {
+            examples.push(Example::Chat(ChatExample { input, output }));
+            continue;
+        }
+
+        // Try parsing `output` as `JsonInferenceOutput` (for JsonExample)
+        if let Ok(output) = serde_json::from_str::<JsonInferenceOutput>(&raw_example.output) {
+            examples.push(Example::Json(JsonExample { input, output }));
+            continue;
+        }
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:430]: We know the function type so we don't have to do this.

File: gateway/src/variant/dicl.rs [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+    pub k: u32, // k as in k-nearest neighbors
+    pub model: String,
+    pub system_instructions: Option<PathBuf>,
+    pub temperature: Option<f32>,
+    pub max_tokens: Option<u32>,
+    pub seed: Option<u32>,
+    #[serde(default)]
+    pub json_mode: JsonMode,
+}
+
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:55]: This file could use a lot more comments, including inline comments.

File: gateway/tests/e2e/dicl.rs [hunk 8]
```diff
@@ -0,0 +1,1133 @@
...
+use tokio::time::sleep;
+use uuid::Uuid;
+
+use crate::common::{
+    get_clickhouse, get_gateway_endpoint, select_chat_inference_clickhouse,
+    select_json_inference_clickhouse, select_model_inferences_clickhouse,
+};
+
+#[tokio::test]
+pub async fn test_empty_dicl_inference_request() {
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/dicl.rs:24]: Document what "empty" means, or rephrase.

File: gateway/tests/e2e/dicl.rs [hunk 9]
```diff
@@ -0,0 +1,1133 @@
...
+    let content_block_type = content_block.get("type").unwrap().as_str().unwrap();
+    assert_eq!(content_block_type, "text");
+    let clickhouse_content = content_block.get("text").unwrap().as_str().unwrap();
+    assert_eq!(clickhouse_content, content);
+
+    let tool_params = result.get("tool_params").unwrap().as_str().unwrap();
+    assert!(tool_params.is_empty());
+
+    let inference_params = result.get("inference_params").unwrap().as_str().unwrap();
+    println!("inference_params: {}", inference_params);
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/dicl.rs:131]: needed?

File: gateway/tests/e2e/dicl.rs [hunk 10]
```diff
@@ -0,0 +1,1133 @@
...
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/dicl.rs:494]: needed?

File: gateway/tests/e2e/dicl.rs [hunk 11]
```diff
@@ -0,0 +1,1133 @@
...
+    let content_block_type = content_block.get("type").unwrap().as_str().unwrap();
+    assert_eq!(content_block_type, "text");
+    let clickhouse_content = content_block.get("text").unwrap().as_str().unwrap();
+    assert_eq!(clickhouse_content, full_content);
+
+    let tool_params = result.get("tool_params").unwrap().as_str().unwrap();
+    assert!(tool_params.is_empty());
+
+    let inference_params = result.get("inference_params").unwrap().as_str().unwrap();
+    println!("inference_params: {}", inference_params);
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/dicl.rs:722]: same

File: gateway/tests/e2e/dicl.rs [hunk 12]
```diff
@@ -0,0 +1,1133 @@
...
+    let json_output = serde_json::from_str::<JsonInferenceOutput>(
+        result.get("output").unwrap().as_str().unwrap(),
+    )
+    .unwrap();
+    let clickhouse_answer = json_output.parsed.unwrap();
+    let clickhouse_answer = clickhouse_answer.get("answer").unwrap().as_str().unwrap();
+    assert_eq!(clickhouse_answer, answer);
+
+    let inference_params = result.get("inference_params").unwrap().as_str().unwrap();
+    println!("inference_params: {}", inference_params);
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/dicl.rs:1031]: same

**Review by GabrielBianconi — CHANGES_REQUESTED** [review]
Needs examples/recipes to merge.

File: gateway/src/config_parser.rs [hunk 13]
```diff
@@ -168,7 +171,13 @@ impl<'c> Config<'c> {
 
         // Validate each function
         for (function_name, function) in &self.functions {
-            function.validate(&self.tools, &self.models, &self.templates, function_name)?;
+            function.validate(
+                &self.tools,
+                &self.models,
+                &self.embedding_models,
+                &self.templates,
```
[the remark this answers is not in the document]
Comment (virajmehta) [reply gateway/src/config_parser.rs:178]: sure, agreed

**Review by GabrielBianconi — APPROVED** [review]
See my changes.

**Pull request merged**